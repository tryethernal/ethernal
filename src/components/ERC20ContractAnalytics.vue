<template>
    <v-container fluid>
        <v-row>
            <v-col cols="12" md="6">
                <v-card outlined class="px-1">
                    <v-card-subtitle>Transfer Volume (14 Days) - UTC Time</v-card-subtitle>
                    <Line-Chart v-if="!transferVolumeLoading && transferVolume.data" :xLabels="transferVolume.xLabels" :data="transferVolume.data" :tooltipUnit="'transfer'" :index="0" />
                    <v-skeleton-loader v-else type="image" class="pa-2"></v-skeleton-loader>
                </v-card>
            </v-col>

            <v-col cols="12" md="6">
                <v-card outlined class="px-1">
                    <v-card-subtitle>Circulating Supply (14 Days) - UTC Time</v-card-subtitle>
                    <Line-Chart v-if="!cumulativeSupplyLoading && cumulativeSupply.data" :xLabels="cumulativeSupply.xLabels" :data="cumulativeSupply.data" :tokenSymbol="tokenSymbol" :index="0" />
                    <v-skeleton-loader v-else type="image" class="pa-2"></v-skeleton-loader>
                </v-card>
            </v-col>

            <v-col cols="12" md="6">
                <v-card outlined class="px-1">
                    <v-card-subtitle>Holder Count (14 Days) - UTC Time</v-card-subtitle>
                    <Line-Chart v-if="!holderHistoryLoading && holderHistory.data" :xLabels="holderHistory.xLabels" :data="holderHistory.data" :tooltipUnit="'holder'" :index="0" />
                    <v-skeleton-loader v-else type="image" class="pa-2"></v-skeleton-loader>
                </v-card>
            </v-col>
        </v-row>
    </v-container>
</template>

<script>
const moment = require('moment');
const ethers = require('ethers');
import LineChart from './LineChart';

export default {
    name: 'ERC20ContractAnalytics',
    props: ['address', 'tokenDecimals', 'tokenSymbol'],
    components: {
        LineChart
    },
    data: () => ({
        transferVolumeLoading: true,
        cumulativeSupplyLoading: true,
        holderHistoryLoading: true,
        transferVolume: {},
        cumulativeSupply: {},
        holderHistory: {}
    }),
    mounted() {
    },
    methods: {
        from: ethers.BigNumber.from,
        formatUnits: ethers.utils.formatUnits,
        moment: moment
    },
    watch: {
        address: {
            immediate: true,
            handler(address) {
                const date14daysAgo = moment().subtract(14, 'days').format('YYYY-MM-DD');
                const dateNow = moment().format('YYYY-MM-DD');

                this.server.getTokenTransferVolume(address, date14daysAgo, dateNow)
                    .then(({ data }) => {
                        this.transferVolume = {
                            xLabels: data.map(t => moment(t.timestamp).format('DD/MM')),
                            data: data.map(t => parseInt(t.count))
                        }
                    })
                    .catch(console.log)
                    .finally(() => this.transferVolumeLoading = false);

                this.server.getTokenCumulativeSupply(address, date14daysAgo, dateNow)
                    .then(({ data }) => {
                        this.cumulativeSupply = {
                            xLabels: data.map(t => moment(t.timestamp).format('DD/MM')),
                            data: data.map(t => parseFloat(this.formatUnits(this.from(t.cumulativeSupply), this.tokenDecimals)))
                        }
                    })
                    .catch(console.log)
                    .finally(() => this.cumulativeSupplyLoading = false);

                this.server.getTokenHolderHistory(address, date14daysAgo, dateNow)
                    .then(({ data }) => {
                        this.holderHistory = {
                            xLabels: data.map(t => moment(t.timestamp).format('DD/MM')),
                            data: data.map(t => parseInt(t.count))
                        }
                    })
                    .catch(console.log)
                    .finally(() => this.holderHistoryLoading = false);
            }
        }
    },
}
</script>
