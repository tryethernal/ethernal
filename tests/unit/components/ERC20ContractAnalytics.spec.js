import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';
import ERC20ContractAnalytics from '@/components/ERC20ContractAnalytics.vue';

let helper;
const stubs = [
    'Line-Chart',
];

describe('ERC20ContractAnalytics.vue', () => {

    beforeEach(() => {
        jest.clearAllMocks();
        helper = new MockHelper();
    });

    it('Should display contract analytics', async () => {
        jest.spyOn(helper.mocks.server, 'getTokenTransferVolume')
            .mockResolvedValueOnce({ data: [
                { timestamp: 0, count: 2 },
                { timestamp: 1, count: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getTokenCumulativeSupply')
            .mockResolvedValueOnce({ data: [
                { timestamp: 0, cumulativeSupply: 2 },
                { timestamp: 1, cumulativeSupply: 2 }
            ]});

        jest.spyOn(helper.mocks.server, 'getTokenHolderHistory')
            .mockResolvedValueOnce({ data: [
                { timestamp: 0, count: 2 },
                { timestamp: 1, count: 2 }
            ]});

        const wrapper = helper.mountFn(ERC20ContractAnalytics, {
            propsData: {
                address: '0x123',
                tokenDecimals: 1,
                tokenSymbol: 'ETL'
            },
            stubs:Â stubs
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
});
