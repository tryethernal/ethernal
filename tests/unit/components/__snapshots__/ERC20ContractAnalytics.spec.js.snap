// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ERC20ContractAnalytics.vue Should display contract analytics 1`] = `
<div class="container container--fluid">
  <div class="row">
    <div class="col-md-6 col-12">
      <div class="px-1 v-card v-sheet v-sheet--outlined theme--light">
        <div class="v-card__subtitle">Transfer Volume (14 Days) - UTC Time</div>
        <line-chart-stub tooltipunit="transfer" data="2,2" xlabels="01/01,01/01" index="0"></line-chart-stub>
      </div>
    </div>
    <div class="col-md-6 col-12">
      <div class="px-1 v-card v-sheet v-sheet--outlined theme--light">
        <div class="v-card__subtitle">Circulating Supply (14 Days) - UTC Time</div>
        <line-chart-stub tokensymbol="ETL" data="0.2,0.2" xlabels="01/01,01/01" index="0"></line-chart-stub>
      </div>
    </div>
    <div class="col-md-6 col-12">
      <div class="px-1 v-card v-sheet v-sheet--outlined theme--light">
        <div class="v-card__subtitle">Holder Count (14 Days) - UTC Time</div>
        <line-chart-stub tooltipunit="holder" data="2,2" xlabels="01/01,01/01" index="0"></line-chart-stub>
      </div>
    </div>
  </div>
</div>
`;
